### PR TITLE
[KYUUBI #4540] Support to deny some users to make connection

### PR DESCRIPTION
**Pull request by zhuyaogai**
### _Why are the changes needed?_

Improvement: https://github.com/apache/kyuubi/issues/4540

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/master/contributing/code/testing.html#running-tests) locally before make a pull request

### _Was this patch authored or co-authored using generative AI tooling?_

No

### CODE REVIEW

**Comment by zhuyaogai** [comment]
@turboFei hi, could you please do a code review and provide some suggestions? Thank you very much!

File: docs/configuration/settings.md [hunk 1]
```diff
@@ -392,6 +392,7 @@ You can configure the Kyuubi properties in `$KYUUBI_HOME/conf/kyuubi-defaults.co
...
 | kyuubi.server.limit.connections.per.user.ipaddress       | &lt;undefined&gt; | Maximum kyuubi server connections per user:ipaddress combination. Any user-ipaddress exceeding this limit will not be allowed to connect.                                                                                                      | int      | 1.6.0 |
 | kyuubi.server.limit.connections.user.unlimited.list                         || The maximum connections of the user in the white list will not be limited.                                                                                                                                                                     | seq      | 1.7.0 |
+| kyuubi.server.limit.connections.user.blacklist                              || The user in the blacklist will be refused to connect to kyuubi server.                                                                                                                                                                         | seq      | 1.8.0 |
```
Comment (turboFei) [inline docs/configuration/settings.md:395]: kyuubi.server.limit.connections.user.limited.list

Comment (turboFei) [reply docs/configuration/settings.md:395]: The user in the limited list will be refused to connect to kyuubi server.

**Comment by turboFei** [comment]
thanks for raising this pr, please help rename all the `blacklist` to `limited.list` to align with the config `kyuubi.server.limit.connections.user.unlimited.list`.

**Comment by zhuyaogai** [comment]
> thanks for raising this pr, please help rename all the `blacklist` to `limited.list` to align with the config `kyuubi.server.limit.connections.user.unlimited.list`.

done.

**Comment by turboFei** [comment]
```
error file=/home/runner/work/kyuubi/kyuubi/kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/opt/AdminCommandLine.scala message=File line length exceeds 100 characters line=111
```

**Comment by zhuyaogai** [comment]
> ```
> error file=/home/runner/work/kyuubi/kyuubi/kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/opt/AdminCommandLine.scala message=File line length exceeds 100 characters line=111
> ```

@turboFei hi, could you tell me what commands can be used for automatic code formatting? For example: `mvn spotless:apply`?

**Comment by pan3793** [comment]
> ... what commands can be used for automatic code formatting?

`dev/reformat`

https://kyuubi.readthedocs.io/en/master/contributing/code/developer.html#format-all-code

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala [hunk 2]
```diff
@@ -361,13 +368,29 @@ class KyuubiSessionManager private (name: String) extends SessionManager(name) {
     batchLimiter.foreach(SessionLimiter.resetUnlimitedUsers(_, unlimitedUsers))
   }
 
+  private[kyuubi] def getLimitedUsers: Set[String] = {
+    limiter.orElse(batchLimiter).map(SessionLimiter.getLimitedUsers).getOrElse(Set.empty)
+  }
+
+  private[kyuubi] def refreshLimitedUsers(conf: KyuubiConf): Unit = {
+    val limitedUsers = conf.get(SERVER_LIMIT_CONNECTIONS_USER_LIMITED_LIST).toSet
```
Comment (bowenliang123) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala:376]: The user list should be filtered ensuring non-empty elements.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/SessionLimiter.scala [hunk 3]
```diff
@@ -101,14 +101,21 @@ class SessionLimiterImpl(userLimit: Int, ipAddressLimit: Int, userIpAddressLimit
...
     userIpAddressLimit: Int,
-    var unlimitedUsers: Set[String])
+    var unlimitedUsers: Set[String],
+    var limitedUsers: Set[String])
   extends SessionLimiterImpl(userLimit, ipAddressLimit, userIpAddressLimit) {
   override def increment(userIpAddress: UserIpAddress): Unit = {
-    if (!unlimitedUsers.contains(userIpAddress.user)) {
+    val user = userIpAddress.user
+    if (StringUtils.isNotBlank(user) && limitedUsers.contains(user)) {
+      val errorMsg = s"Connection refused because the user is limited. (user: $user)"
```
Comment (bowenliang123) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/SessionLimiter.scala:114]: Misleading message. This is not part of the connection handshake.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/SessionLimiter.scala [hunk 4]
```diff
@@ -101,14 +101,21 @@ class SessionLimiterImpl(userLimit: Int, ipAddressLimit: Int, userIpAddressLimit
...
     ipAddressLimit: Int,
     userIpAddressLimit: Int,
-    var unlimitedUsers: Set[String])
+    var unlimitedUsers: Set[String],
+    var limitedUsers: Set[String])
   extends SessionLimiterImpl(userLimit, ipAddressLimit, userIpAddressLimit) {
   override def increment(userIpAddress: UserIpAddress): Unit = {
-    if (!unlimitedUsers.contains(userIpAddress.user)) {
+    val user = userIpAddress.user
+    if (StringUtils.isNotBlank(user) && limitedUsers.contains(user)) {
```
Comment (bowenliang123) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/SessionLimiter.scala:113]: I'm not familiar with the usage of SessionLimiter. Is this the right way or timing to do the rejection here?  cc @wForget

File: kyuubi-server/src/test/scala/org/apache/kyuubi/server/api/v1/AdminResourceSuite.scala [hunk 5]
```diff
@@ -125,6 +125,19 @@ class AdminResourceSuite extends KyuubiFunSuite with RestFrontendTestHelper {
     assert(200 == response.getStatus)
   }
 
+  test("refresh limited users of the kyuubi server") {
+    var response = webTarget.path("api/v1/admin/refresh/limited_users")
+      .request()
+      .post(null)
+    assert(405 == response.getStatus)
```
Comment (bowenliang123) [inline kyuubi-server/src/test/scala/org/apache/kyuubi/server/api/v1/AdminResourceSuite.scala:132]: `403 Forbidden` may be a more suitable response code than `405 Method not allowed`.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala (hunk 2 above)
Comment (zhuyaogai) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionManager.scala]: done.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/SessionLimiter.scala (hunk 3 above)
Comment (zhuyaogai) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/session/SessionLimiter.scala]: done.

File: kyuubi-server/src/test/scala/org/apache/kyuubi/server/api/v1/AdminResourceSuite.scala (hunk 5 above)
Comment (zhuyaogai) [reply kyuubi-server/src/test/scala/org/apache/kyuubi/server/api/v1/AdminResourceSuite.scala:132]: I wrote the code based on other test cases. If you think the 403 status code should be better, I can modify it (including all other test cases).

**Comment by turboFei** [comment]
could you check the log error and address it？

**Comment by zhuyaogai** [comment]
> could you check the log error and address it？
Sorry, I thought this was not my issue, so I didn't pay attention to it. It has now been fixed.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 6]
```diff
@@ -2614,6 +2614,15 @@ object KyuubiConf {
       .toSequence()
       .createWithDefault(Nil)
 
+  val SERVER_LIMIT_CONNECTIONS_USER_LIMITED_LIST: ConfigEntry[Seq[String]] =
+    buildConf("kyuubi.server.limit.connections.user.limited.list")
+      .doc("The user in the limited list will be limited to connect to kyuubi server.")
```
[the remark this answers is not in the document]
Comment (zhuyaogai) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:2619]: @pan3793 hi, do you mean that the user is configured in both `user.unlimited.list` and `user.limited.list`? In my code logic, it seems that the priority of the latter is higher.

[the remark this answers is not in the document]
Comment (pan3793) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:2619]: yes. and the logic makes sense, but we should document it clear

[the remark this answers is not in the document]
Comment (pan3793) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:2619]: let's say there is a user name "tom", what if `user.unlimited.list` and `user.limited.list` is configured but neither includes "tom"?

[the remark this answers is not in the document]
Comment (zhuyaogai) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala]: > the logic makes sense, but we should document it clear

done.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/SessionLimiter.scala (hunk 4 above)
Comment (pan3793) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/session/SessionLimiter.scala:113]: I have the same question, the implementation looks incorrect, limited does not mean rejection.

Comment (zhuyaogai) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/session/SessionLimiter.scala:113]: @pan3793 hi, I am new to kyuubi, could you give me more details about the meaning of `limited user`? From the issue  context I cannot get more input.

Comment (pan3793) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/session/SessionLimiter.scala:113]: I suppose when "limited user" is configured, connections from those users will be limited not rejected.

e.g. 

```
kyuubi.server.limit.connections.user.limited.list=tom,jerry
kyuubi.server.limit.connections.per.user=10
```

user "tom" and "jerry" are limited to opening up to 10 connections for each user, other users are not limited.

Comment (pan3793) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/session/SessionLimiter.scala:113]: 1. that's why we discuss the priority of `limited.user.list` and `unlimited.user.list`, I think it's fine that `limited.user.list` takes high priority. (it means when `limited.user.list` is configured, ignore `unlimited.user.list`)
2. `user-ipaddress` should be affected by this change too. and we can introduce a similar change for `ipaddress` in another PR.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 7]
```diff
@@ -2614,6 +2614,17 @@ object KyuubiConf {
...
 
+  val SERVER_LIMIT_CONNECTIONS_USER_LIMITED_LIST: ConfigEntry[Seq[String]] =
+    buildConf("kyuubi.server.limit.connections.user.limited.list")
+      .doc("The user in the limited list will be limited to connect to kyuubi server, " +
+        "if the user has configured both user.unlimited.list and user.limited.list, " +
+        "the priority of the latter is higher.")
+      .version("1.8.0")
+      .serverOnly
+      .stringConf
+      .toSequence()
```
Comment (bowenliang123) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:2625]:
```suggestion
      .toSet()
```
With #5185 introduced set collection support for configs, you could rebase to the master branch and use it.

Comment (zhuyaogai) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:2625]: Thank you for your reminder.

**Review by turboFei — APPROVED** [review]

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala (hunk 7 above)
Comment (turboFei) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:2625]: LGTM after this comment addressed

Comment (zhuyaogai) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:2625]: @turboFei hi, do you mean that aside from the above case, my current code implementation is okay? but @pan3793  thinks it's incorrect.

Comment (pan3793) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:2625]: I had an offline discussion with @turboFei, and we reached a consensus that:

1. The current implementation is **deny** semantic, which is a good feature but **limited** is not a properly name
2. You could either change the configuration name to `...deny.user.list`, or keep the `...limited.user.list` and change the implementation as I commented above https://github.com/apache/kyuubi/pull/5182#discussion_r1304545505, I think the latter approach is simple, and we can implement the **deny** in another PR.

Comment (zhuyaogai) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala]: @pan3793 hi, I have made some changes to the code, and I hope I understood what you wanted. Could you please review the code and give me some suggestions? Thank you!

Comment (pan3793) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala]: @zhuyaogai will take a look tomorrow, have a good weekend :)

Comment (zhuyaogai) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala]: @pan3793 BTW, I found that is there a problem with the current master branch code? When some users open some sessions and before closing the sessions, this user is added to the `unlimited.list`, then the counter related to this user will not be cleared.

Comment (zhuyaogai) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala]: I think we don't need to check any limit in `decrement` method, and let the already opened sessions decrement limit counter. Please correct me if I am wrong.

Comment (bowenliang123) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala]: > I had an offline discussion with @turboFei, and we reached a consensus that:
> 
> 1. The current implementation is **deny** semantic, which is a good feature but **limited** is not a properly name
> 2. You could either change the configuration name to `...deny.user.list`, or keep the `...limited.user.list` and change the implementation as I commented above [[KYUUBI #4540] [Improvement] Support to limit some user to make connection. #5182 (comment)](https://github.com/apache/kyuubi/pull/5182#discussion_r1304545505), I think the latter approach is simple, and we can implement the **deny** in another PR.

I have similar concerns in these variations. I was confused by the **deny** and **limited**.
As for me, a limited user list is not the contrast to an unlimited user, and only a feature for  deny users list is necessary.
What's more I would like to see when the server reject user connection , this feature should be part of authentication ,such as User filerter,  but not an injected mechanism in such Limiter. Or did I miss any information for better understanding?

I would suggest, 
1. Add detailed and summarised design and key points to this PR description, as a simple link to the original issue is not enough
2. Add constructive discussion in proposed change and the reason for possible approaches

I'm fine if we accept this feature, especially when some of us have a mutual understanding of enough information.

**Comment by zhuyaogai** [comment]
@pan3793 hi, will we have further discussions?

**Comment by pan3793** [comment]
@zhuyaogai I'm not sure if we have gaps in the logic design. The code implementation and the test cases do not cover the combinations of `limited.users` and `unlimited.users`.

1. Neither `limited.users` nor `unlimited.users` are configured
2. Only `limited.users` is configured
3. Only `unlimited.users` is configured
4. Both `limited.users` and `unlimited.users` are configured

seems the current code only considers case 4?

> that's why we discuss the priority of `limited.user.list` and `unlimited.user.list`, I think it's fine that `limited.user.list` takes high priority. (it means when `limited.user.list` is configured, ignore `unlimited.user.list`)

And I suppose we reached a consensus on what the behavior should be like for case 4, but the code looks do not match.

**Review by turboFei — APPROVED** [review]
LGTM, thanks for the efforts

**Review by pan3793 — APPROVED** [review]
LGTM, thanks for your patience and great work! @zhuyaogai

**Review by bowenliang123 — APPROVED** [review]

**Comment by pan3793** [comment]
Thanks, merged to master

**Pull request closed without merging**